File: api/compounds.py
from flask import Blueprint, jsonify, request
from services.compound_service import CompoundService
from config import Config

compounds_bp = Blueprint('compounds', __name__)
compound_service = CompoundService()

@compounds_bp.route('/compounds', methods=['GET'])
def get_compounds():
    """
    获取化合物列表
    
    Query Parameters:
        - page: 页码（默认1）
        - page_size: 每页数量（默认20）
        - type: 化合物类型
        - search: 搜索关键词
        - sort_by: 排序字段
        - sort_order: 排序方向（asc/desc）
    """
    try:
        # 获取请求参数
        page = request.args.get('page', 1, type=int)
        page_size = request.args.get('page_size', Config.DEFAULT_PAGE_SIZE, type=int)
        compound_type = request.args.get('type')
        search = request.args.get('search')
        sort_by = request.args.get('sort_by', 'global_id')
        sort_order = request.args.get('sort_order', 'asc')
        
        # 验证参数
        page = max(1, page)
        page_size = min(max(1, page_size), Config.MAX_PAGE_SIZE)
        
        # 获取数据
        result = compound_service.get_compounds_list(
            page=page,
            page_size=page_size,
            compound_type=compound_type,
            search=search,
            sort_by=sort_by,
            sort_order=sort_order
        )
        
        return jsonify({
            'status': 'success',
            'data': result
        })
        
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': str(e)
        }), 500

@compounds_bp.route('/compounds/<int:compound_id>', methods=['GET'])
def get_compound_detail(compound_id):
    """获取化合物详情"""
    try:
        compound = compound_service.get_compound_detail(compound_id)
        
        if compound:
            return jsonify({
                'status': 'success',
                'data': compound
            })
        else:
            return jsonify({
                'status': 'error',
                'message': 'Compound not found'
            }), 404
            
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': str(e)
        }), 500

@compounds_bp.route('/compounds/search', methods=['POST'])
def search_compounds():
    """搜索化合物"""
    try:
        data = request.get_json()
        query = data.get('query', '')
        search_type = data.get('type', 'all')
        
        results = compound_service.search_compounds(query, search_type)
        
        return jsonify({
            'status': 'success',
            'data': results,
            'total': len(results)
        })
        
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': str(e)
        }), 500

@compounds_bp.route('/compounds/statistics', methods=['GET'])
def get_statistics():
    """获取统计信息"""
    try:
        stats = compound_service.get_statistics()
        
        return jsonify({
            'status': 'success',
            'data': stats
        })
        
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': str(e)
        }), 500

from services.target_service import TargetService

target_service = TargetService()

@compounds_bp.route('/compounds/<int:compound_id>/targets', methods=['GET'])
def get_compound_targets(compound_id):
    """获取化合物的预测靶点"""
    try:
        # 获取分页参数
        page = request.args.get('page', 1, type=int)
        page_size = request.args.get('page_size', 10, type=int)
        
        # 先获取化合物信息
        compound = compound_service.get_compound_detail(compound_id)
        
        if not compound:
            return jsonify({
                'status': 'error',
                'message': 'Compound not found'
            }), 404
        
        # 获取所有靶点信息（不带分页参数）
        targets_data = target_service.get_compound_targets(
            compound['compound_type'],
            compound['id']
        )
        
        # 在Python中实现分页
        all_targets = targets_data.get('targets', [])
        total_targets = len(all_targets)
        
        # 计算分页
        start_idx = (page - 1) * page_size
        end_idx = start_idx + page_size
        paginated_targets = all_targets[start_idx:end_idx]
        
        # 构建分页信息
        total_pages = (total_targets + page_size - 1) // page_size
        pagination = {
            'page': page,
            'page_size': page_size,
            'total': total_targets,
            'total_pages': total_pages,
            'has_next': page < total_pages,
            'has_prev': page > 1
        }
        
        return jsonify({
            'status': 'success',
            'data': {
                'compound_info': compound,
                'targets': paginated_targets,
                'statistics': targets_data.get('statistics', {}),
                'pagination': pagination
            }
        })
        
    except Exception as e:
        return jsonify({
            'status': 'error',
            'message': str(e)
        }), 500

File: static/js/compounds.js
/**
 * Compounds Module JavaScript - Handles list and detail page functionality
 */

let compoundsTable = null;

// Initialize after page load
document.addEventListener('DOMContentLoaded', function() {
    // Initialize appropriate functionality based on current page
    if (document.getElementById('compoundsTable')) {
        initializeCompoundsListPage();
    }
    
    if (window.compoundData) {
        initializeCompoundDetailPage();
    }
});

/**
 * Initialize compounds list page
 */
function initializeCompoundsListPage() {
    initializeDataTable();
    bindListPageEvents();
    
    // Set initial filter values
    if (window.pageConfig) {
        if (window.pageConfig.compoundType && window.pageConfig.compoundType !== 'all') {
            document.getElementById('typeFilter').value = window.pageConfig.compoundType;
        }
        if (window.pageConfig.searchQuery) {
            document.getElementById('searchInput').value = window.pageConfig.searchQuery;
        }
    }
}

/**
 * Initialize DataTables with fixed pagination
 */
function initializeDataTable() {
    compoundsTable = $('#compoundsTable').DataTable({
        processing: true,
        serverSide: true,
        ajax: {
            // Fixed: Use the correct endpoint that returns DataTables format
            url: '/compounds/api/list',
            type: 'GET',
            data: function(d) {
                // Add custom filter parameters to DataTables request
                d.compound_type = document.getElementById('typeFilter').value !== 'all' ? 
                                 document.getElementById('typeFilter').value : null;
                d.search_custom = document.getElementById('searchInput').value || null;
                return d;
            },
            error: function(xhr, error, thrown) {
                console.error('DataTable error:', error);
                Utils.showToast('Failed to load data', 'error');
            }
        },
        columns: [
            {
                data: 'global_id',
                title: 'ID',
                width: '80px',
                render: function(data) {
                    return `<strong>${data}</strong>`;
                }
            },
            {
                data: 'chinese_name',
                title: 'Chinese Name',
                render: function(data) {
                    return data || '<span class="text-muted">Unnamed</span>';
                }
            },
            {
                data: 'Name',
                title: 'English Name',
                render: function(data) {
                    return data || '<span class="text-muted">Unnamed</span>';
                }
            },
            {
                data: 'compound_type',
                title: 'Type',
                width: '120px',
                render: function(data) {
                    const types = {
                        '挥发油': { label: 'Essential Oils', color: 'success' },
                        '三萜': { label: 'Triterpenes', color: 'info' },
                        '甾醇': { label: 'Sterols', color: 'warning' }
                    };
                    const type = types[data] || { label: data, color: 'secondary' };
                    return `<span class="badge bg-${type.color}">${type.label}</span>`;
                }
            },
            {
                data: 'Molecular_Formula',
                title: 'Molecular Formula',
                render: function(data) {
                    return data ? `<code>${data}</code>` : '<span class="text-muted">N/A</span>';
                }
            },
            {
                data: 'Molecular_Weight',
                title: 'Molecular Weight',
                width: '140px',
                render: function(data) {
                    return data ? `${parseFloat(data).toFixed(2)} g/mol` : '<span class="text-muted">N/A</span>';
                }
            },
            {
                data: null,
                title: 'Actions',
                orderable: false,
                width: '100px',
                render: function(data, type, row) {
                    return `
                        <div class="btn-group btn-group-sm">
                            <button class="btn btn-outline-primary" onclick="viewCompound(${row.global_id})" title="View Details">
                                <i class="fas fa-eye"></i>
                            </button>
                            <button class="btn btn-outline-success" onclick="viewTargets(${row.global_id})" title="Predicted Targets">
                                <i class="fas fa-bullseye"></i>
                            </button>
                        </div>
                    `;
                }
            }
        ],
        pageLength: 20,
        lengthMenu: [[10, 20, 50, 100], [10, 20, 50, 100]],
        language: {
            url: '//cdn.datatables.net/plug-ins/1.13.7/i18n/en.json',
            processing: "Loading compounds...",
            search: "Search compounds:",
            lengthMenu: "Show _MENU_ compounds per page",
            info: "Showing _START_ to _END_ of _TOTAL_ compounds",
            infoEmpty: "No compounds found",
            infoFiltered: "(filtered from _MAX_ total compounds)",
            paginate: {
                first: "First",
                last: "Last",
                next: "Next",
                previous: "Previous"
            }
        },
        order: [[0, 'asc']],
        responsive: true
    });
}

/**
 * Bind list page events
 */
function bindListPageEvents() {
    // Filter changes - reload table when type filter changes
    document.getElementById('typeFilter').addEventListener('change', function() {
        compoundsTable.ajax.reload();
    });
    
    // Search input enter key
    document.getElementById('searchInput').addEventListener('keypress', function(e) {
        if (e.key === 'Enter') {
            compoundsTable.ajax.reload();
        }
    });
    
    // Real-time search (debounced)
    let searchTimeout;
    document.getElementById('searchInput').addEventListener('input', function() {
        clearTimeout(searchTimeout);
        searchTimeout = setTimeout(() => {
            compoundsTable.ajax.reload();
        }, 500);
    });
}

/**
 * View compound details
 */
function viewCompound(compoundId) {
    window.location.href = `/compounds/${compoundId}`;
}

/**
 * View predicted targets
 */
function viewTargets(compoundId) {
    window.location.href = `/compounds/${compoundId}#targets`;
}

/**
 * Initialize compound detail page
 */
function initializeCompoundDetailPage() {
    loadTargetsData();
}

/**
 * Load 2D molecular structure
 */
function loadMoleculeStructure() {
    const viewer = document.getElementById('moleculeViewer');
    if (!viewer || !window.compoundData || !window.compoundData.SMILES) {
        viewer.innerHTML = '<p class="text-center text-muted">Unable to load structure: Missing SMILES data.</p>';
        return;
    }

    // Create a Canvas element
    const canvas = document.createElement('canvas');
    canvas.id = 'smiles-canvas';
    viewer.innerHTML = ''; // Clear placeholder
    viewer.appendChild(canvas);

    try {
        // Initialize SmilesDrawer
        const drawer = new SmilesDrawer.Drawer({
            width: viewer.clientWidth || 400,
            height: 350,
            bondThickness: 2,
            bondLength: 15
        });

        // Draw structure
        SmilesDrawer.parse(window.compoundData.SMILES, function(tree) {
            drawer.draw(tree, 'smiles-canvas', 'light', false);
        }, function(err) {
            console.error("SMILES parsing error:", err);
            viewer.innerHTML = '<p class="text-center text-danger">Failed to parse SMILES structure.</p>';
        });
    } catch (error) {
        console.error("SmilesDrawer error:", error);
        viewer.innerHTML = '<p class="text-center text-warning">Molecular visualization unavailable.</p>';
    }
}

/**
 * Show full InChI
 */
function showFullInChI() {
    if (window.compoundData && window.compoundData.InChI) {
        document.getElementById('fullInchiText').value = window.compoundData.InChI;
        const inchiModal = new bootstrap.Modal(document.getElementById('inchiModal'));
        inchiModal.show();
    }
}

/**
 * Load predicted targets data with pagination
 */
async function loadTargetsData(page = 1, pageSize = 10, sortBy = 'score') {
    const container = document.getElementById('targetsContainer');
    
    if (!container || !window.compoundData) return;
    
    try {
        // Show loading state
        container.innerHTML = `
            <div class="text-center py-4">
                <div class="spinner-border text-primary" role="status">
                    <span class="visually-hidden">Loading...</span>
                </div>
                <p class="mt-2 text-muted">Loading predicted targets...</p>
            </div>
        `;
        
        // Fetch paginated data from API
        const response = await apiClient.get(`/compounds/${window.compoundData.global_id}/targets`, {
            page: page,
            page_size: pageSize,
            sort_by: sortBy
        });
        
        const targets = response.data.targets || [];
        const statistics = response.data.statistics || {};
        const pagination = response.data.pagination || {};
        
        displayTargets(targets, statistics, pagination, page, pageSize);
        
    } catch (error) {
        console.error('Error loading targets:', error);
        container.innerHTML = `
            <div class="alert alert-danger">
                <i class="fas fa-exclamation-triangle me-2"></i>
                Failed to load predicted targets
            </div>
        `;
    }
}

/**
 * Display predicted targets with pagination
 */
function displayTargets(targets, statistics, pagination, currentPage, pageSize) {
    const container = document.getElementById('targetsContainer');
    
    if (!targets || targets.length === 0) {
        container.innerHTML = `
            <div class="text-center text-muted py-4">
                <i class="fas fa-info-circle fa-2x mb-3"></i>
                <p>No predicted targets available</p>
            </div>
        `;
        return;
    }
    
    // Sort targets by score (highest first)
    const sortedTargets = targets.sort((a, b) => (b.score || 0) - (a.score || 0));
    
    // Build targets list HTML
    let targetsHtml = `
        <div class="mb-3">
            <div class="d-flex justify-content-between align-items-center">
                <small class="text-muted">
                    Showing ${((currentPage - 1) * pageSize) + 1}-${Math.min(currentPage * pageSize, pagination.total || 0)} 
                    of ${pagination.total || 0} targets
                </small>
                <div class="btn-group btn-group-sm">
                    <button class="btn btn-outline-secondary btn-sm" onclick="sortTargets('score')" title="Sort by Score">
                        <i class="fas fa-sort-numeric-down"></i>
                    </button>
                    <button class="btn btn-outline-secondary btn-sm" onclick="sortTargets('name')" title="Sort by Name">
                        <i class="fas fa-sort-alpha-down"></i>
                    </button>
                </div>
            </div>
        </div>
    `;
    
    targetsHtml += sortedTargets.map(target => {
        const score = target.score || 0;
        const scoreClass = score >= 0.95 ? 'success' : score >= 0.90 ? 'warning' : 'secondary';
        const geneName = target.gene_name || target.From || 'Unknown';
        const geneSymbol = target.gene_symbol || '';
        
        return `
            <div class="border-bottom pb-2 mb-2 target-item">
                <div class="d-flex justify-content-between align-items-start">
                    <div class="flex-grow-1">
                        <h6 class="mb-1">
                            <a href="/targets/${encodeURIComponent(geneSymbol || geneName)}" 
                               class="text-decoration-none" target="_blank" title="View target details">
                                ${geneName}
                            </a>
                        </h6>
                        ${geneSymbol && geneSymbol !== geneName ? 
                          `<small class="text-muted">Gene Symbol: <code>${geneSymbol}</code></small>` : ''}
                    </div>
                    <div class="text-end">
                        <span class="badge bg-${scoreClass}" title="Prediction Score">
                            ${score.toFixed(4)}
                        </span>
                    </div>
                </div>
            </div>
        `;
    }).join('');
    
    // Add pagination controls if needed
    if (pagination.total_pages > 1) {
        targetsHtml += createPaginationHTML(pagination, currentPage);
    }
    
    container.innerHTML = targetsHtml;
}

/**
 * Create pagination HTML
 */
function createPaginationHTML(pagination, currentPage) {
    const totalPages = pagination.total_pages || 1;
    let paginationHtml = `
        <nav aria-label="Targets pagination" class="mt-3">
            <ul class="pagination pagination-sm justify-content-center">
    `;
    
    // Previous button
    paginationHtml += `
        <li class="page-item ${currentPage <= 1 ? 'disabled' : ''}">
            <button class="page-link" onclick="loadTargetsData(${currentPage - 1})" 
                    ${currentPage <= 1 ? 'disabled' : ''} title="Previous page">
                <i class="fas fa-chevron-left"></i>
            </button>
        </li>
    `;
    
    // Page numbers
    const startPage = Math.max(1, currentPage - 2);
    const endPage = Math.min(totalPages, currentPage + 2);
    
    if (startPage > 1) {
        paginationHtml += `
            <li class="page-item">
                <button class="page-link" onclick="loadTargetsData(1)">1</button>
            </li>
        `;
        if (startPage > 2) {
            paginationHtml += `<li class="page-item disabled"><span class="page-link">...</span></li>`;
        }
    }
    
    for (let i = startPage; i <= endPage; i++) {
        paginationHtml += `
            <li class="page-item ${i === currentPage ? 'active' : ''}">
                <button class="page-link" onclick="loadTargetsData(${i})">${i}</button>
            </li>
        `;
    }
    
    if (endPage < totalPages) {
        if (endPage < totalPages - 1) {
            paginationHtml += `<li class="page-item disabled"><span class="page-link">...</span></li>`;
        }
        paginationHtml += `
            <li class="page-item">
                <button class="page-link" onclick="loadTargetsData(${totalPages})">${totalPages}</button>
            </li>
        `;
    }
    
    // Next button
    paginationHtml += `
        <li class="page-item ${currentPage >= totalPages ? 'disabled' : ''}">
            <button class="page-link" onclick="loadTargetsData(${currentPage + 1})" 
                    ${currentPage >= totalPages ? 'disabled' : ''} title="Next page">
                <i class="fas fa-chevron-right"></i>
            </button>
        </li>
    `;
    
    paginationHtml += `
            </ul>
        </nav>
    `;
    
    return paginationHtml;
}

/**
 * Sort targets
 */
function sortTargets(sortBy) {
    // Reload current page data with sort parameter
    loadTargetsData(1, 10, sortBy);
}

/**
 * Copy to clipboard
 */
function copyToClipboard(text) {
    navigator.clipboard.writeText(text).then(() => {
        Utils.showToast('Copied to clipboard', 'success');
    }).catch(err => {
        console.error('Copy failed:', err);
        Utils.showToast('Copy failed', 'error');
    });
}